source/server/server.cc

```diff
@@ -95,6 +95,17 @@ InstanceImpl::~InstanceImpl() {
   // Stop logging to file before all the AccessLogManager and its dependencies are
   // destructed to avoid crashing at shutdown.
   file_logger_.reset();
+
+  // Destruct the ListenerManager explicitly, before InstanceImpl's local init_manager_ is
+  // destructed.
+  //
+  // The ListenerManager's DestinationPortsMap contains FilterChainSharedPtrs. There is a rare race
```

Comment:
I'm not sure how this is a RARE race, can you elaborate a bit more? Particularly, when RdsRouteConfigSubscription get registered to top-level InitManager?
- FilterChains contains HttpConnectionManager is pretty common
- HttpConnectionManager contains RdsRouteConfigProvider is pretty common too
- Any dynamic RdsRouteConfigProvider contains a RdsRouteConfigSubscription

Reply:
I think the circumstances are common but the resulting segfault is rare. 

I'm not 100% on how memory works here under the hood, but my personal theory is that when the subscription goes to unregister itself from the list of `targets_` inside `InitManagerImpl`, `targets_` has just been freed but not not overwritten (?) so this operation often succeeds. @mrice32 might have more color here.

Reply:
Yes use-after-free is always rare without ASAN, does that mean you can have a config which always trigger use-after-free with ASAN?

Reply:
That's an interesting thought, the answer is surely yes. I'll do some digging and report back.

Reply:
Once you have that, we could have some sort of test here. Could be an integration test with shutdown or sth similar.